[mlir][VectorOps] Support string literals in `vector.print`

MacDue opened the pull request:
Printing strings within integration tests is currently quite annoyingly
verbose, and can't be tucked into shared helpers as the types depend on
the length of the string:

```
llvm.mlir.global internal constant @hello_world("Hello, World!\0")

func.func @entry() {
  %0 = llvm.mlir.addressof @hello_world : !llvm.ptr<array<14 x i8>>
  %1 = llvm.mlir.constant(0 : index) : i64
  %2 = llvm.getelementptr %0[%1, %1]
    : (!llvm.ptr<array<14 x i8>>, i64, i64) -> !llvm.ptr<i8>
  llvm.call @printCString(%2) : (!llvm.ptr<i8>) -> ()
  return
}
```

So this patch adds a simple extension to `vector.print` to simplify
this:
```
func.func @entry() {
   // Print a vector of characters ;)
   vector.print str "Hello, World!"
   return
}
```

Most of the logic for this is now shared with `cf.assert` which already
does something similar.

Depends on #68694

MacDue committed:
[mlir][VectorOps] Support string literals in `vector.print`

Printing strings within integration tests is currently quite annoyingly
verbose, and can't be tucked into shared helpers as the types depend on
the length of the string:

```
llvm.mlir.global internal constant @hello_world("Hello, World!\0")

func.func @entry() {
  %0 = llvm.mlir.addressof @hello_world : !llvm.ptr<array<14 x i8>>
  %1 = llvm.mlir.constant(0 : index) : i64
  %2 = llvm.getelementptr %0[%1, %1]
    : (!llvm.ptr<array<14 x i8>>, i64, i64) -> !llvm.ptr<i8>
  llvm.call @printCString(%2) : (!llvm.ptr<i8>) -> ()
  return
}
``

So this patch adds a simple extension to `vector.print` to simplify
this:
```
func.func @entry() {
   // Print a vector of characters ;)
   vector.print str "Hello, World!"
   return
}
```

Most of the logic for this is now shared with `cf.assert` which already
does something similar.

MacDue commented:
cc @c-rhodes

c-rhodes commented on mlir/test/Integration/Dialect/Vector/CPU/test-hello-world.mlir:
`test-print-str.mlir` for consistency with the other print tests

c-rhodes reviewed (commented):
thanks for the patch Ben! I think this is generally a really useful feature and will clean up some of the tests nicely

MacDue committed:
Fixups

MacDue commented:
Split the test updates to: https://github.com/llvm/llvm-project/pull/68973

c-rhodes reviewed (approved):
LGTM thanks Ben. Would be good to have some input from others so please allow time for that before landing, cheers.

banach-space commented on mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:
```diff
@@ -2476,12 +2477,18 @@ def Vector_TransposeOp :
-  Vector_Op<"print", []>,
+  Vector_Op<"print", [
+    PredOpTrait<
+      "`source` or `punctuation` are not set when printing strings",
```
Perhaps I am a bit pedantic, but this reads a bit off :) Perhaps "Strings mixed with `source` and `punctations` are not allowed/supported" ?

MacDue replied:
When this constraint fails it outputs:
```
op failed to verify `source` or `punctuation` are not set when printing strings
```
If I switch it, that does not really work :disappointed: 
```
op failed to verify strings mixed with source and punctations are not allowed/supported
```

banach-space commented on mlir/lib/Conversion/LLVMCommon/PrintCallHelper.cpp:
```diff
@@ -0,0 +1,64 @@
+  return uniqueName;
+}
+
+void mlir::LLVM::createPrintStrCall(OpBuilder &builder, Location loc,
```
Is this simply `createPrintMsg` after moving?

MacDue replied:
Pretty much, I changed it to `Str` as message felt a little specific

banach-space reviewed (approved):
LGTM, thanks!

aartbik reviewed (commented):
Thanks for adding this very useful feature!

Note that calling "puts" puts a direct dependence on the std lib for this to work, while originally the idea of the vector print support in the CRunnerlibrary was that users could provide implementation of "just" a few basic print operations (printI64, printU64, etc.) and then provide their own implementation, usually indeed using a std lib, as in

extern "C" void printNewline() { fputc('\n', stdout); }


So from that sense, would it make sense to provide

extern "C" void printString(char *) { fputs(s, stdout); }

and use that extra indirection to stay along the original philosophy I had for this library?

MacDue committed:
Use `printCStr()` rather than `puts()`

PrintCallHelper is the only use of this, so we can safely switch.

MacDue commented:
> Thanks for adding this very useful feature!
> 
> Note that calling "puts" puts a direct dependence on the std lib for this to work, while originally the idea of the vector print support in the CRunnerlibrary was that users could provide implementation of "just" a few basic print operations (printI64, printU64, etc.) and then provide their own implementation, usually indeed using a std lib, as in
> 
> extern "C" void printNewline() { fputc('\n', stdout); }
> 
> So from that sense, would it make sense to provide
> 
> extern "C" void printString(char *) { fputs(s, stdout); }
> 
> and use that extra indirection to stay along the original philosophy I had for this library?

I've switched this to use `printCStr()` rather than `puts()` directly now, as that already exists in the runner library :+1:

MacDue committed:
Fixup test checks and naming

aartbik commented:
Apologies for nitpicking... but, printCString belongs to the group of memref prints, defined in RunnerUtils.h. The group of methods that constitute the "small set of primitives" that are easy to port to other platforms live in CRunnerUtils.h, and are grouped together here:

//===----------------------------------------------------------------------===//
// Small runtime support library for vector.print lowering during codegen.
//===----------------------------------------------------------------------===//
extern "C" MLIR_CRUNNERUTILS_EXPORT void printI64(int64_t i);
extern "C" MLIR_CRUNNERUTILS_EXPORT void printU64(uint64_t u);
extern "C" MLIR_CRUNNERUTILS_EXPORT void printF32(float f);
extern "C" MLIR_CRUNNERUTILS_EXPORT void printF64(double d);
extern "C" MLIR_CRUNNERUTILS_EXPORT void printOpen();
extern "C" MLIR_CRUNNERUTILS_EXPORT void printClose();
extern "C" MLIR_CRUNNERUTILS_EXPORT void printComma();
extern "C" MLIR_CRUNNERUTILS_EXPORT void printNewline();

It feels like we need to add a printString(char*s) to this set (and then also the implementation at the top of CRunnerUtils.cpp).

In the big picture, I realize this may seem like too much nitpicking, but it it is much closer to the design philosophy of the vector.print support when I first added it (enabling the first actual running end-to-end MLIR tests ;-)

MacDue committed:
Add printString to CRunnerUtils and use that instead

I've also defined this in RunnerUtils so linking either or both gives
you printString, this avoids the need to update a bunch of tests that
use cf.assert.

MacDue commented:
> Apologies for nitpicking... but, printCString belongs to the group of memref prints, defined in RunnerUtils.h. The group of methods that constitute the "small set of primitives" that are easy to port to other platforms live in CRunnerUtils.h, and are grouped together here:
> 
> //===----------------------------------------------------------------------===// // Small runtime support library for vector.print lowering during codegen. //===----------------------------------------------------------------------===// extern "C" MLIR_CRUNNERUTILS_EXPORT void printI64(int64_t i); extern "C" MLIR_CRUNNERUTILS_EXPORT void printU64(uint64_t u); extern "C" MLIR_CRUNNERUTILS_EXPORT void printF32(float f); extern "C" MLIR_CRUNNERUTILS_EXPORT void printF64(double d); extern "C" MLIR_CRUNNERUTILS_EXPORT void printOpen(); extern "C" MLIR_CRUNNERUTILS_EXPORT void printClose(); extern "C" MLIR_CRUNNERUTILS_EXPORT void printComma(); extern "C" MLIR_CRUNNERUTILS_EXPORT void printNewline();
> 
> It feels like we need to add a printString(char*s) to this set (and then also the implementation at the top of CRunnerUtils.cpp).
> 
> In the big picture, I realize this may seem like too much nitpicking, but it it is much closer to the design philosophy of the vector.print support when I first added it (enabling the first actual running end-to-end MLIR tests ;-)

No worries! I've added printString to CRunnerUtils and switched to that. I've also weakly defined in implementation in RunnerUtils since this is shared with cf.assert and several tests there link RunnerUtils, but not CRunnerUtils, so it's easier if linking either gives you printString().

banach-space commented:
> I've also defined an implementation in RunnerUtils since this is shared with `cf.assert

Hi Ben, I think that we should avoid defining the same symbol both in `RunnerUtils` and `CRunnerUtils`. I know that the linker won't complain (IIRC, I will just use the first definition that it encounters). However, once the definitions in two libraries diverge, people might see different behavior depending on the order in which they list these runtime libraries. That would be undesirable. 

Perhaps we should just focus on `vector.print` and skip `cf.assert` for now? Would that simplify things?

MacDue committed:
Use puts for cf.assert and printString for vector.print

MacDue commented:
> Perhaps we should just focus on vector.print and skip cf.assert for now? Would that simplify things?

We can't skip `cf.assert` as this is shared code, but I've now reworked things so you can specify the runtime function to use. `cf.assert` will now keep using `puts()` and `vector.print` will use a new `printString()` function (which is defined only in CRunnerUtils). 

---

`printCString()` should be removed in a later patch (and uses updated to use `vector.print str` or `printString()` directly)

banach-space commented on mlir/lib/ExecutionEngine/RunnerUtils.cpp:
```diff
@@ -158,7 +158,8 @@ extern "C" void printMemrefC64(int64_t rank, void *ptr) {
 }
 
-extern "C" void printCString(char *str) { printf("%s", str); }
+/// Deprecated. This should be unified with printString from CRunnerUtils.
+extern "C" void printCString(char *str) { fputs(str, stdout); }
```
[nit] I would just skip these changes altoghether.

banach-space reviewed (approved):
Thanks Ben, I believe that this addresses Aart's concerns about not depending on stdlib and also for `vector.print` to be fully customizable via hooks in CRunnerUtils.cpp.

Please wait for @aartbik to confirm before landing :)

aartbik reviewed (approved):
Thank you very kindly for taking my suggestion very seriously. Looks great!

jroelofs commented on mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:
```diff
@@ -2477,12 +2478,18 @@ def Vector_TransposeOp :
+  Vector_Op<"print", [
+    PredOpTrait<
+      "`source` or `punctuation` are not set when printing strings",
+      CPred<"!getStringLiteral() || (!getSource() && getPunctuation() == PrintPunctuation::NewLine)">
```
Why isn't other punctuation allowed? It would be very useful to be able to print strings without the trailing newline.

c-rhodes replied:
Strings don't have a trailing newline.

jroelofs replied:
The lowering for them in `LLVM::createPrintStrCall` adds one unconditionally.

jroelofs replied:
I'll post a complete patch for this in a bit, but I was thinking something along the lines of:
```
diff --git a/mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp b/mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
index b66b55ae8d57..fd5aec1982eb 100644
--- a/mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
+++ b/mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
@@ -1531,8 +1531,28 @@ public:
 
     auto punct = printOp.getPunctuation();
     if (auto stringLiteral = printOp.getStringLiteral()) {
+      std::string str;
+      llvm::raw_string_ostream punctuatedLiteral(str);
+      if (punct == PrintPunctuation::Open)
+        punctuatedLiteral << "( ";
+      punctuatedLiteral << stringLiteral->str();
+      switch (punct) {
+      case PrintPunctuation::Close:
+        punctuatedLiteral << " )";
+        break;
+      case PrintPunctuation::Comma:
+        punctuatedLiteral << ", ";
+        break;
+      case PrintPunctuation::NewLine:
+        punctuatedLiteral << '\n';
+        break;
+      case PrintPunctuation::Open:
+      case PrintPunctuation::NoPunctuation:
+        break;
+      }
       LLVM::createPrintStrCall(rewriter, loc, parent, "vector_print_str",
-                               *stringLiteral, *getTypeConverter());
+                               punctuatedLiteral.str(), *getTypeConverter(),
+                               /*addNewLine=*/false);
     } else if (punct != PrintPunctuation::NoPunctuation) {
       emitCall(rewriter, printOp->getLoc(), [&] {
         switch (punct) {
```

so that:
```
diff --git a/mlir/test/Integration/Dialect/Vector/CPU/test-print-str.mlir b/mlir/test/Integration/Dialect/Vector/CPU/test-print-str.mlir
index 78d6609ccaf9..b47c5b38f783 100644
--- a/mlir/test/Integration/Dialect/Vector/CPU/test-print-str.mlir
+++ b/mlir/test/Integration/Dialect/Vector/CPU/test-print-str.mlir
@@ -8,6 +8,16 @@
 func.func @entry() {
    // CHECK: Hello, World!
    vector.print str "Hello, World!"
+
+   // CHECK-NEXT: Nice to meet you ( finally, today, and in this place )
+   vector.print str "Nice to meet you " punctuation <no_punctuation>
+   vector.print str "finally" punctuation <open>
+   vector.print punctuation <comma>
+   vector.print str "today" punctuation <comma>
+   vector.print str "and in " punctuation <no_punctuation>
+   vector.print str "this place" punctuation <close>
+   vector.print punctuation <newline>
+
    // CHECK-NEXT: Bye!
    vector.print str "Bye!"
    return
```

c-rhodes replied:
hm that's odd, I verified no newline is printed for strings yesterday and that seems to be the case, are we seeing different behaviour? The `createPrintStrCall` for `vector.print <str>` already has `addNewLine=false`
https://github.com/llvm/llvm-project/blob/4014e2e045f5160ce9cbb9562d151f540d61c0bb/mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp#L1520-L1523

although for empty string it will print newline as that wont be entered.
